### PR TITLE
Fix indentation when using braceless block with single Defn

### PR DESCRIPTION
**Pull request by tgodzik**
Previously, it was not possible to have a single val in a block and not have braces, which is why it didn't work for Scala 3.

Now, we make sure that indentation is added always if we have a single statement, which is a Defn.

### CODE REVIEW

**Comment by tgodzik** [comment]
I think this is the last issue I found, once we merge this, we can release scalameta and release RC3

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 1]
```diff
@@ -1985,7 +1985,8 @@ class FormatOps(
 
     trait Impl {
       def tryGetSplits(ft: FormatToken, nft: FormatToken)(implicit
-          style: ScalafmtConfig
+          style: ScalafmtConfig,
+          line: sourcecode.Line
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:1989]: why do you need to pass the line? and if you really do, can that be a separate commit, with item description?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 2]
```diff
@@ -2081,15 +2086,20 @@ class FormatOps(
...
       ): Option[Seq[Split]] = {
         val leftOwner = ft.meta.leftOwner
         findTreeWithParentSimple(nft.meta.rightOwner)(_ eq leftOwner) match {
           case Some(t: Term.Block)
-              if t.stats.lengthCompare(1) > 0 && isBlockStart(t, nft) =>
+              if (t.stats.lengthCompare(1) > 0 ||
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2097]: you don't need to check length anymore. just remove this line and replace`exists` with `forall` on next.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 3]
```diff
@@ -2081,15 +2086,20 @@ class FormatOps(
...
+                getBlockSingleStat(t).exists(_.isNot[Term]))
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2098]: is it isNot[term], or is[Defn]?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 4]
```diff
@@ -2081,15 +2086,20 @@ class FormatOps(
...
+                && isBlockStart(t, nft) =>
             Some(getSplitsMaybeBlock(ft, nft, t, true))
-          case _ => None
+          case other =>
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2101]: must be debug leftover. could you please restore two lines.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 5]
```diff
@@ -1999,7 +2000,10 @@ class FormatOps(
     // if  match  return  then  throw  try  while  yield
     def unapply(
         ftMeta: FormatToken.Meta
-    )(implicit style: ScalafmtConfig): Option[Seq[Split]] = {
+    )(implicit
+        style: ScalafmtConfig,
+        line: sourcecode.Line
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2005]: this line is always the same, right? it's not useful for debugging.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 6]
```diff
@@ -2081,15 +2086,20 @@ class FormatOps(
     }
 
     object BlockImpl extends Impl {
+
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2089]: this empty line makes it inconsistent with other Impl implementations?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala [hunk 7]
```diff
@@ -634,7 +634,7 @@ object TreeOps {
     case _ => false
   }
 
-  def getTermSingleStat(t: Term): Option[Tree] =
+  def getTermSingleStat(t: Tree): Option[Tree] =
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala:637]: please rename to getTree

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 8]
```diff
@@ -2367,7 +2388,9 @@ class FormatOps(
         allowMain: Boolean
     )(implicit line: sourcecode.Line, style: ScalafmtConfig): Seq[Split] = {
       val multiStat = isTreeMultiStatBlock(tree)
-      val forceNL = multiStat || shouldBreakInOptionalBraces(nft)
+      val singleNoTerms = getTermSingleStat(tree).exists(_.isNot[Term])
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2391]: same logic as above. remove multistat, change here exists to forall. also, the check inside repeats, should we create a helper function to share with line 2098 above?

**Review by kitbellew — CHANGES_REQUESTED** [review]

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 1 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:1989]: I spend a bunch of time trying to figure out why we were adding a newline in line 2084 of Router and it turned out it wasn't in in at all. Maybe we could have the while line plus source file? I can move it to a separate commit.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 6 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2089]: Ach sorry, that's left over from debugging. Got thrown off by the checks not running and haven't checked for silly changes

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 3 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2098]: I think that is safer, but Defn should suffice

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 1 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:1989]: yeah, i know, I'm used to it by now. your idea of passing the filename is interesting, though, perhaps we should do that later.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 3 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2098]: there's a similar logic in RedundantBraces (to avoid removing them), perhaps we should have a shared check.

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2098]: and if it's safer, let's keep it. you should know better anyway. i still haven't figured out the difference between tree and term.

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2098]: > there's a similar logic in RedundantBraces (to avoid removing them), perhaps we should have a shared check.

that one checks for Defn or Term.Function (`innerOk`). somewhat different, not much to share. but perhaps we should change `innerOk` to do `tree.isNot[Term] || tree.is[Term.Function]` (similar to this proposal).

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 1 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:1989]: I wonder how to do it and not have to change every single:
```
implicit  line: sourcecode.Line
```

Into
```
implicit  line: sourcecode.Line,
implicit  file: sourcecode.File
```

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 2 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2097]: Wait but if I dontt check the length then stats with a single Term will also get caught in this case, no ?

Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2097]: We want either a block with many terms or a block with a single non term. It seems to me exclusive.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 1 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: I removed all the added `line`

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 2 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Changed it  all a bit, let me know what you think. I think we still need the compare the length.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 3 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Changed the logic in RedundantBraces a bit too.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 4 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: fixed

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 5 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Removed

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 6 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Removef

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala (hunk 7 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala]: Renamed, as well as the method below.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 8 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Created a helper function.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala [hunk 9]
```diff
@@ -634,19 +634,25 @@ object TreeOps {
...
+  def getTreeLineSpan(b: Tree): Int =
     if (b.tokens.isEmpty) 0
     else {
       val pos = b.pos
       pos.endLine - pos.startLine
     }
 
+  def hasBlockOnlyNonTermStats(t: Tree) =
+    t match {
+      case b: Term.Block => b.stats.forall(_.isNot[Term])
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala:652]: we never really wanted to check all terms, right? just one and only.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 1 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: https://github.com/scalameta/scalafmt/pull/2514 <- quick search and replace

**Review by kitbellew — CHANGES_REQUESTED** [review]

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 10]
```diff
@@ -2087,7 +2087,8 @@ class FormatOps(
         val leftOwner = ft.meta.leftOwner
         findTreeWithParentSimple(nft.meta.rightOwner)(_ eq leftOwner) match {
           case Some(t: Term.Block)
-              if t.stats.lengthCompare(1) > 0 && isBlockStart(t, nft) =>
+              if (t.stats.lengthCompare(1) > 0 || hasBlockOnlyNonTermStats(t))
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:2090]: Och, I missed it is a None in case of multiple stats. Thanks!

[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Fixed!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala [hunk 11]
```diff
@@ -634,19 +634,25 @@ object TreeOps {
...
 
-  def getTermLineSpan(b: Tree): Int =
+  def getTreeLineSpan(b: Tree): Int =
     if (b.tokens.isEmpty) 0
     else {
       val pos = b.pos
       pos.endLine - pos.startLine
     }
 
+  def hasBlockOnlyNonTermStats(t: Tree) =
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala]: Changed!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 12]
```diff
@@ -2367,7 +2368,9 @@ class FormatOps(
         allowMain: Boolean
     )(implicit line: sourcecode.Line, style: ScalafmtConfig): Seq[Split] = {
       val multiStat = isTreeMultiStatBlock(tree)
-      val forceNL = multiStat || shouldBreakInOptionalBraces(nft)
+      val singleNoTerms = hasBlockOnlyNonTermStats(tree)
+      val forceNL =
+        multiStat || singleNoTerms || shouldBreakInOptionalBraces(nft)
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Replaced

**Review by kitbellew — APPROVED** [review]
